src/lib.rs

```diff
@@ -0,0 +1,127 @@
+#[deny(clippy::all)]
+///! Graph API Traits
+///
+use rand::SeedableRng;
+
+#[allow(dead_code)]
+/// A graph layer name.
+struct Layer(&'static str);
+
+/// Abstract object in a graph, eg. node or edge.
+trait GraphObject {
+    /// Identifier of all graph objects.
+    type Id;
+
+    /// Local graph object data.
+    type Data;
+
+    /// Return the edge id.
+    fn id(&self) -> Self::Id;
+
+    /// Get local data.
+    fn data(&self) -> &Self::Data;
+
+    /// Get mutable local data.
+    fn data_mut(&self) -> &mut Self::Data;
+}
+
+/// A graph  node.
+trait Node: GraphObject {
+    /// The edge type connected to this node.
+    type Edge: GraphObject;
+
+    /// Get connected nodes.
+    fn neighbors(&self) -> Iterator<Item = &Self>;
+
+    /// Get incoming and outgoing edges.
+    fn edges(&self) -> Iterator<Item = &Self::Edge>;
+}
+
+/// A graph edge between two nodes.
+trait Edge: GraphObject {
+    /// Get the edge weight.
+    fn weight(&self) -> f64;
+
+    /// The object from which the edge starts.
+    fn from(&self) -> <Self as GraphObject>::Id;
+
+    /// The object to which the edge points.
+    fn to(&self) -> <Self as GraphObject>::Id;
+}
+
+/// The Graph API
+trait GraphAPI<G: Graph> {
+    /// Entropy used to seed the CPRNG.
+    type Entropy;
+
+    /// Add a graph layer.
+    fn add_layer(&mut self, layer: Layer);
+
+    /// Remove a graph layer. The layer must not contain any nodes.
+    fn remove_layer(&mut self, layer: &Layer);
+
+    /// Return an immutable graph of the given layer.
+    fn graph(&self, layer: &Layer) -> Option<&G>;
+
+    /// Return the mutable graph of the given layer.
+    fn graph_mut(&mut self, layer: &Layer) -> Option<&mut G>;
```

Comment:
Scary 😬

Reply:
You're going to have to modify the graph somehow, no?

Reply:
Of course! the bigger question is whether or not such mutation should be visible at the API level, i.e. if users of the Graph API should be given the power of mutating the graph themselves vs having an API function that takes a bunch of operations on the graph and does the mutation internally. Dunno yet which is best, I am brainstorming here!

Reply:
What I mean is, even if you implement a high-level interface for doing this, eventually you will have to apply the mutations to the graph, and hence will need a mutable reference to it.

Reply:
Yes, we are on the same page on this 😉

Reply:
Perhaps we could have an interface that doesn't let you add/remove nodes, but just mutate the data on them and the edges. Then it'll be up to the implementation of the Graph API to do the dirty work. But someone's gotta do it ;)